File: 56_EntryNodeOfLoop.py

# -*- coding:utf-8 -*-
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def EntryNodeOfLoop(self, pHead):
        # write code here
        if pHead is None or pHead.next is None:
            return None
        p_slow = pHead
        p_fast = pHead
        while p_fast and p_fast.next:
            p_slow = p_slow.next
            p_fast = p_fast.next.next
            if p_slow == p_fast:
                p_slow = pHead
                while p_slow != p_fast:
                    p_slow = p_slow.next
                    p_fast = p_fast.next
                return p_slow
        return None

 
# l: not in loop length 
# r: loop length
# t: meet length in loop 
# n: num loop 

# l + nr + t = 2(l + t)
# l + t = nr 
# once meet in the loop:
# p_slow go back to pHead
# p_fast continue but move one step every time
# they will meet again

# since the first time meeting steps (l + t) 
# is the loop node numbers

# equal to: p_fast first move the loop numbers steps
#     	  then move with s_slow one step each time
#     	  they will meet again

File: 56_EntryNodeOfLoop.cpp
/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
        val(x), next(NULL) {
    }
};
*/
class Solution {
public:
    ListNode* EntryNodeOfLoop(ListNode* pHead)
    {
        if(pHead == NULL) return NULL;
        ListNode *fast = pHead;
        ListNode *slow = pHead;
        if(fast -> next != NULL) fast = (fast -> next) -> next;
        slow = slow -> next;
        while(fast != slow && fast -> next != NULL)
        {
            fast = (fast -> next) -> next;
            slow = slow -> next;
        }
        
        if(fast == NULL || fast -> next == NULL) return NULL;
        slow = pHead;
        while(fast != slow)
        {
            fast = fast -> next;
            slow = slow -> next;
        }
        return fast;
    }
};

// 运行时间：3ms
// 占用内存：476k

// 给一个链表，若其中包含环，请找出该链表的环的入口结点，否则，输出null。
//    *meet

// |-----|
// |     |
// |-----|
// | t
// |

// before entrance length = t;
// round length = r;
// entrance to meet point length = l;
 
// t + r + l = 2 * (t + l)
// r = t + l

// len(meet to entrance) = len(start to entrance)

// use fast and slow pointer to find the meet point 

// use two pointers, 
// one from meet to entrance
// one from start to entrance 
// both step = 1
// to find the entrance point